.github/workflows/reusable-unit-split.yml

```
@@ -0,0 +1,140 @@
+name: reusable-unit-split
+
+on:
+  workflow_call:
+    inputs:
+      directory:
+        required: true
+        type: string
+      go-version:
+        required: true
+        type: string
+      go-arch:
+        required: false
+        type: string
+        default: ""
+      uploaded-binary-name:
+        required: false
+        type: string
+        default: "consul-bin" 
+      runs-on:
+        description: An expression indicating which kind of runners to use.
+        required: false
+        type: string
+        default: '["custom", "linux", "xl"]'
+      args:
+        required: false
+        type: string
+        default: ""
+      runner-count: 
+        required: false
+        type: number
+        default: 1
+      go-test-flags:
+        required: false
+        type: string
+        default: ""
+
+env:
+  TEST_RESULTS: /tmp/test-results
+  GOTESTSUM_VERSION: 1.8.2
+  GOARCH: ${{inputs.go-arch}}
+  TOTAL_RUNNERS: ${{inputs.runner-count}}
+  
+jobs:
+  set-test-package-matrix:
+    runs-on: ubuntu-latest
+    outputs:
+      package-matrix: ${{ steps.set-matrix.outputs.matrix }}
+    steps:
+      - uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+      - uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568 # pin@v3.5.0
+        with:
+          go-version-file: 'go.mod'
+      - id: set-matrix
+        run: ./.github/scripts/set_test_package_matrix.sh ${{env.TOTAL_RUNNERS}}
+
+  go-test:
+    runs-on: ${{ fromJSON(inputs.runs-on) }}
+    name: "go-test"
+    needs:
+      - set-test-package-matrix
+    strategy:
+      fail-fast: false
+      matrix:
+        package: ${{ fromJson(needs.set-test-package-matrix.outputs.package-matrix) }}
+    steps:   
+      - name: ulimit
+        run: |
+          echo "Soft limits"
+          ulimit -Sa
+          echo "Hard limits"
+          ulimit -Ha   
+      - uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # pin@v3.3.0
+      - uses: actions/setup-go@6edd4406fa81c3da01a34fa6f6343087c207a568 # pin@v3.5.0
+        with:
+          go-version-file: 'go.mod'
+      - name: Setup go mod cache
+        uses: actions/cache@v3
+        with:
+          path: |
+            ~/.cache/go-build
+            ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
```

You might be able to just do this? https://github.com/actions/setup-go#caching-dependency-files-and-build-outputs
```suggestion
          go-version-file: 'go.mod'
          cache: true
```